algorithms/geospatial_query.hpp

```diff
@@ -0,0 +1,180 @@
+                {
+                    use_directions = checkSegmentBearing(data, bearing, bearing_range);
+                    if (use_directions.first || use_directions.second)
+                    {
+                        has_big_component = has_big_component || !data.component.is_tiny;
+                        has_small_component = has_small_component || data.component.is_tiny;
+                    }
+                }
+
+                return use_directions;
+            },
+            [&has_big_component](const std::size_t num_results, const float)
+            {
+                return num_results > 0 && has_big_component;
+            });
+
+
+        if (results.size() == 0)
+        {
+            return std::make_pair(PhantomNode {}, PhantomNode {});
+        }
+
+        BOOST_ASSERT(results.size() > 0);
+        return std::make_pair(MakePhantomNode(input_coordinate, results.front()).second, MakePhantomNode(input_coordinate, results.back()).second);
+    }
+
+  private:
+    inline std::vector<std::pair<double, PhantomNode>>
+    MakePhantomNodes(const FixedPointCoordinate &input_coordinate,
+                     const std::vector<EdgeData> &results) const
+    {
+        std::vector<std::pair<double, PhantomNode>> distance_and_phantoms(results.size());
+        std::transform(results.begin(), results.end(), distance_and_phantoms.begin(),
+                       [this, &input_coordinate](const EdgeData &data)
+                       {
+                           return MakePhantomNode(input_coordinate, data);
+                       });
+        return distance_and_phantoms;
+    }
+
+    inline std::pair<double, PhantomNode>
+    MakePhantomNode(const FixedPointCoordinate &input_coordinate, const EdgeData &data) const
+    {
+        FixedPointCoordinate point_on_segment;
+        float ratio;
+        const auto current_perpendicular_distance = coordinate_calculation::perpendicular_distance(
+            coordinates->at(data.u), coordinates->at(data.v), input_coordinate, point_on_segment,
+            ratio);
+
+        auto transformed =
+            std::make_pair(current_perpendicular_distance, PhantomNode{data, point_on_segment});
+
+        ratio = std::min(1.f, std::max(0.f, ratio));
+
+        if (SPECIAL_NODEID != transformed.second.forward_node_id)
+        {
+            transformed.second.forward_weight *= ratio;
+        }
+        if (SPECIAL_NODEID != transformed.second.reverse_node_id)
+        {
+            transformed.second.reverse_weight *= 1.f - ratio;
+        }
+        return transformed;
+    }
+
+    inline std::pair<bool, bool> checkSegmentBearing(const EdgeData &segment,
+                                                     const float filter_bearing,
+                                                     const float filter_bearing_range)
+    {
+        const float forward_edge_bearing =
+            coordinate_calculation::bearing(coordinates->at(segment.u), coordinates->at(segment.v));
```

Comment:
Why the `.at` member function here instead of the subscript operator `[]`? We are sure that it's there, no need to use `.at` for bounds checking?